docs/lofi-hifi-backends.md

```diff
@@ -0,0 +1,176 @@
+# Introduce Lofi and Hifi Commands for Key ilab functions to Increase Model Fidelity
+
+
+This document describes adding different data generation, mixing, and training backends to ilab to enable higher fidelity training using the backend code.
+
+
+Currently all training is done via qlora or the like. Adding the following commands will enable higher fidelity training and introduce commands such as data mixing.
+
+
+## Key Component
+
+
+### Building off of the InstructLab Structure Redesign
+
+
+After github.com/instructlab/instructlab/pull/990 is merged, ilab will use a parent -> child command structure. This proposal operates under that new structure.
+
+
+This will be the new structure:
+
+
+```console
+ilab
+|
+|_______model
+|       |
+|       |____convert
+|       |____download
+|       |____train (--convert)
+|       |____serve   (-i)
+|       |____chat
+|       |____inference
+|       |
+|_______data
+|       |
+|       |____generate
+|       |
+|_______config
+|       |
+|       |____init
+|       |
+|_______tax
+|       |
+|       |____diff
+|       |____check
+|       |____download
+```
+
+
+
+
+And this Will be the structure after these new commands are added:
+
+
+
+```console
+ilab
+|
+|_______model
+|       |
+|       |____convert
+|       |____download
+|       |____train (--convert)
+|       |    |
+|       |    |______lofi *
+|       |    |______hifi *
```

Comment:
I'm not a fan of any variation of "low" and "high" in the commands. I think it's too limiting. In practice, I expect we'll need more of a sliding scale over time and not a binary set of choices.

An alternative would be to expose more directly the algorithm we're talking about, and algorithm-specific tunables. 

* `ilab train qlora ...`
* `ilab train deepspeed ...`
* ...

something like that.

It's possible to also have an `ilab train auto` where we try to make some automatic decisions, or maybe even a set of t-shirt size type options geared at a range of resource profiles, but I think it may be premature to rush into those while things still seem to be evolving so much.

Reply:
to add to this, I think that if a user tries to run `ilab train` by itself, they should be shown a help message outlining the choices and outlining hardware limitations and other considerations of each, e.g.
```
usage: ilab train <training_algo>

Available Training Options:
  qlora: Requires n GPUs, m GB RAM | suitable for running on laptop...
  fsdp: Requires n GPUs, m GB RAM | suitable for running on RHEL AI/larger infra...
```

@RobotSail also brought up the brilliant suggestion of having it interactively help the user choose, which I am very much on board with.

Reply:
no longer relevant due to doc overhaul, resolving